### PR TITLE
Implement selective reset as a part of new deploy routine

### PR DESCRIPTION
**Pull request by pyromaniac**

### CODE REVIEW

File: lib/chewy/index/params.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        filter = {ids: {values: [@index.derivable_index_name]}}
+        JSON.parse(storage.filter(filter).first.try!(:value) || '{}')
+      end
+
+      def current
+        @index.params_hash.as_json
+      end
+
+      def changed?
+        current != stored
```
Comment (barthez) [inline lib/chewy/index/params.rb]: The `current` params would have symbol keys, `stored` would have string keys, as they are from `JSON.parse`. How does it work for you?

File: lib/chewy/search/request.rb [hunk 2]
```diff
@@ -26,7 +26,7 @@ class Request
         timeout min_score source stored_fields search_after
         load script_fields suggest aggs aggregations none
         indices_boost rescore highlight total total_count
-        total_entries types delete_all count exists? exist? find
+        total_entries types delete_all count exists? exist? find pluck
```
Comment (barthez) [inline lib/chewy/search/request.rb]: How does it apply to rest of changes?

File: lib/chewy/index/params.rb (hunk 1 above)
Comment (pyromaniac) [inline lib/chewy/index/params.rb]: as_json converts everything to strings, even dates

File: lib/chewy/search/request.rb (hunk 2 above)
Comment (pyromaniac) [inline lib/chewy/search/request.rb]: I've just noticed that pluck was not delegated :) Why not?

Comment (barthez) [inline lib/chewy/search/request.rb]: Make it separate commit then, please.

File: lib/chewy/index/params.rb (hunk 1 above)
Comment (barthez) [inline lib/chewy/index/params.rb]: ah, I was confused. Thanks!

File: lib/chewy/search/request.rb (hunk 2 above)
Comment (pyromaniac) [inline lib/chewy/search/request.rb]: C'mon

File: lib/chewy/index/params.rb (hunk 1 above)
Comment (pyromaniac) [inline lib/chewy/index/params.rb]: Yeah, this is covered in specs, so we will notice if the behavior will change :)

File: lib/chewy/index.rb [hunk 3]
```diff
@@ -177,8 +178,12 @@ def self.mappings_hash
       mappings.present? ? {mappings: mappings} : {}
     end
 
-    def self.index_params
+    def self.specification_hash
```
Comment (barthez) [inline lib/chewy/index.rb]: Deprecated alias? WDYT? Let's be nice to people.

File: lib/chewy/rake_helper.rb [hunk 4]
```diff
@@ -1,33 +1,35 @@
...
-            payload[:errors].each do |action, errors|
-              puts "    #{action.to_s.humanize} errors:"
-              errors.each do |error, documents|
-                puts "      `#{error}`"
-                puts "        on #{documents.count} documents: #{documents}"
-              end
-            end
+    IMPORT_CALLBACK = lambda do |_name, start, finish, _id, payload|
+      duration = (finish - start).round(2)
+      puts "  Imported #{payload[:type]} for #{duration}s, documents total: #{payload[:import].try(:[], :index).to_i}"
```
Comment (barthez) [inline lib/chewy/rake_helper.rb]: So, you have decided we will see stash indexing log in console?
```
Imported Chewy::Stash::Specification for 0.04s, documents total: 1
```

Comment (pyromaniac) [inline lib/chewy/rake_helper.rb]: yup, at least for now, it is easier

File: README.md [hunk 5]
```diff
@@ -908,6 +912,19 @@ rake chewy:update[-users,projects] # updates every index in application except s
 
 `rake chewy:reset` performs zero-downtime reindexing as described [here](https://www.elastic.co/blog/changing-mapping-with-zero-downtime). So basically rake task creates a new index with uniq suffix and then simply aliases it to the common index name. The previous index is deleted afterwards (see `Chewy::Index.reset!` for more details).
 
+#### `chewy:deploy`
+
+This rake task is especially useful during the production deploy. Currently it executes selective reset, this means that an index will be reset only if the index specification (settings or mappings) had been changed, otherwise the reset of this index will be skipped.
```
Comment (barthez) [inline README.md]: > ~had~ has been changed

File: lib/chewy/index/specification.rb [hunk 6]
```diff
@@ -1,26 +1,55 @@
 module Chewy
   class Index
+    # Index specification is a combination if index settings and
```
Comment (barthez) [inline lib/chewy/index/specification.rb]: > combination ~if~ of index

File: lib/chewy/index/specification.rb [hunk 7]
```diff
@@ -1,26 +1,55 @@
...
+    # mappings. The idea behind this class is that specification
+    # can be locked in the `Chewy::Stash` between resets, so it is
+    # possible to track changes. In the future it is planned to
+    # be way smarter but right now `rake chewy:deploy` checks
+    # that there were changes and resets the index only if anything
```
Comment (barthez) [inline lib/chewy/index/specification.rb]: checks ~that~ if there

File: lib/chewy/index/specification.rb [hunk 8]
```diff
@@ -1,26 +1,55 @@
...
+      # empty hash if nothing is stored yet.
+      #
+      # @return [Hash] hash produced with JSON parser
       def locked
         filter = {ids: {values: [@index.derivable_index_name]}}
         JSON.parse(Chewy::Stash::Specification.filter(filter).first.try!(:value) || '{}')
       end
 
+      # Simply returns `Chewy::Index.specification_hash`, but
+      # prepared for JSON with `as_json` method. This means all the
```
Comment (barthez) [inline lib/chewy/index/specification.rb]: keys are strings and there are only values of types handled in JSON.

File: lib/chewy/index/specification.rb [hunk 9]
```diff
@@ -1,26 +1,55 @@
...
+      # keys are strings.
+      #
+      # @see Chewy::Index.specification_hash
+      # @return [Hash] a JSON-ready hash
       def current
         @index.specification_hash.as_json
       end
 
+      # Compares previously locked and current specifications
```
Comment (barthez) [inline lib/chewy/index/specification.rb]: period is missing.
